api/src/main/java/bio/terra/tanagra/service/search/SqlVisitor.java

```diff
@@ -0,0 +1,85 @@
+package bio.terra.tanagra.service.search;
+
+import bio.terra.tanagra.service.search.Expression.AttributeExpression;
+import bio.terra.tanagra.service.search.Filter.ArrayFunction;
+import bio.terra.tanagra.service.search.Filter.BinaryFunction;
+import bio.terra.tanagra.service.search.Filter.Visitor;
+import java.util.stream.Collectors;
+
+/** Visitors for walking query constructs to create SQL. */
+// TODO consider how this may need to be split for different SQL backends.
+public class SqlVisitor {
+  /** A {@link Visitor} for creating SQL for filters. */
+  static class FilterVisitor implements Visitor<String> {
+    private final SearchContext searchContext;
+
+    FilterVisitor(SearchContext searchContext) {
+      this.searchContext = searchContext;
+    }
+
+    @Override
+    public String visitArrayFunction(ArrayFunction arrayFunction) {
```

Comment:
I wish there were something better/safer available than raw SQL strings, but my searches have not been too successful. JPA Criteria API might be the closest thing (also portable across datasets), but I'm very skeptical it will be able to handle all the JOINs you want to construct. Data apps team may also have some experience with trying that and can better describe the limitations.

Reply:
One thing I'm eyeing is jOOQ DSL API, particularly as a "dynamic" schema. https://www.jooq.org/doc/3.16/manual/getting-started/use-cases/jooq-as-a-sql-builder-without-codegeneration/

I'd like to get some basics in place with raw strings, then do try implementing a parallel set of visitors for jOOQ and see how that goes.